File: First-Edition/BookSamples/CH08/src/Cellular/Assets/GameOfLife.hs
{-# LANGUAGE DataKinds      #-}
{-# LANGUAGE GADTs          #-}
{-# LANGUAGE KindSignatures #-}
{-# LANGUAGE TypeFamilies   #-}
module Cellular.Assets.GameOfLife where

import Cellular.Language.Automaton
import Cellular.Implementation.Automaton

import GHC.TypeLits


type A = State "Alive" 1
type D = State "Dead" 0

type Neighbors3  = NeighborsCount A '[3  ]
type Neighbors23 = NeighborsCount A '[2,3]

type GoLStep = Step ('DefState D)
  '[ 'StateTransition D A Neighbors3
   , 'StateTransition A A Neighbors23
   ]

type GoLRule = Rule
  "Game of Life"
  "gol"
  (AdjacentsLvl 1)
  GoLStep

File: First-Edition/BookSamples/rosetta-stone/rust/rsc1p3_type_level_interfaces/src/cellular/assets/game_of_life.rs

use crate::automaton::IStateTransition;
use crate::automaton::StateTransition;
use crate::automaton::Step;
use crate::cellular::language::extensions::State;
use crate::cellular::language::extensions::UnknownState;
use crate::cellular::language::extensions::NeighborsCount;
use crate::cellular::language::extensions::AdjacentsLvl;
use crate::cellular::language::extensions::Rule;

use tl_str_macro::tl_str;
use tl_list_lib::tl_list;
use tl_list_lib::tl_i32_list;

use std::marker::PhantomData;

pub type A = State<tl_str!("Alive"), 1>;
pub type D = State<tl_str!("Dead"), 0>;

pub type Neighbors3  = NeighborsCount<A, tl_i32_list![3]>;
pub type Neighbors23 = NeighborsCount<A, tl_i32_list![2, 3]>;

pub type GolTransitions = tl_list![
  IStateTransition,
  StateTransition<D, A, Neighbors3>,
  StateTransition<A, A, Neighbors23>,
  StateTransition<UnknownState, UnknownState, Neighbors3>
  ];

pub type GoLStep = Step<D, GolTransitions>;

pub type GoLRule = Rule<
  tl_str!("Game of Life"),
  tl_str!("gol"),
  AdjacentsLvl<1>,
  GoLStep>;

#[allow(dead_code)]
const GOL_RULE_EVIDENCE: PhantomData::<GoLRule> = PhantomData;

